### ISSUE TITLE
Occasional Lags on playing audio and graphic stimuli

### ISSUE THREAD

**Issue by alirezakzmi** [Psychtoolbox-3/Psychtoolbox-3#455]
Hi all,

I am implementing a behavioral task and only have images and audio stimuli. Everything works fine except some occasional lags. Sometimes, particularly in the middle of the test, the audio stimuli is played in a stretched manner (Assume that my audio stimulus is 'Turtle' with duration of 1 second but it plays something like: 'Tuuuurrrrrtttttleeee' with duration of 2 or 3 seconds). During this stretched audio play I feel some lags on the graphics, too. This is going to happen for some successive stimuli and after some seconds (after some screen flips and Audio stimulus changes) it becomes normal again. Consequently, I feel that an extra CPU load cause this. Maybe I have to delete and recreate some buffers or handles. I already tried to delete and recreate Audio player handle but it causes the Matlab to crash! (This is a reproducible Error!). In the following, I provided some detailed information regarding the way I coded my experiment. I would kindly appreciate it if anyone can help me through this. 

Program Details:
I use Audio Player, Graphic Player, Mouse and Keyboard Polling and GetSecs classes and functions. I created the corresponding objects and keep using them to the end of the experiment. My program is implemented in object-oriented manner.

I use Screen('Flip'), Screen('MakeTexture'), and Screen('DrawTexture') functions to display my images. I create a window with these commands:
`[window, windowRect] =PsychImaging('OpenWindow', Obj.Screen_Prop.Screen_Num,...
                                   Obj.Screen_Prop.Background_Color);            
[screenXpixels, screenYpixels] = Screen('WindowSize', window);
[xCenter, yCenter] = RectCenter(windowRect);
Screen('BlendFunction', window, 'GL_SRC_ALPHA', 'GL_ONE_MINUS_SRC_ALPHA');`

I won't close the window and only flip it. Also, sometimes I use different modes of Flip command to display something while keeping the previous ones on the screen or clear screen using a Flip and display something new.

Besides, I use the audio player of the Psychtoolbox to play my audio stimuli. I initialized the audio player through the following command.
`   InitializePsychSound(1);
    nrchannels = 2;    
    pahandle = PsychPortAudio('Open', [], 1, 1, Fs, nrchannels);`

Also, I tried the following options for loading the Audio player with no difference on avoiding the issue:
`pahandle = PsychPortAudio('Open', [], [], 0, Fs, nrchannels);`

I always reload the Audio buffer with this command to fill it with my different stimuli:
`PsychPortAudio('FillBuffer', Obj.Parameters.Handle, AudioData);`

More than 256 times the Audio buffer is reloaded and more than 300 times I flipped the screen through the test.

I use Matlab 2017a 64bit on Windows 10 Pro 64bit  with 12 Gb of RAM and Intel Core i5-3337U 1.8-2.6GHz. I test another Laptop with better performance and same OS but the issue wasn't solved. I already updated all my Drivers and the PTB to the latest available version.

**Comment by kleinerm** [comment]
Does it happen with our own sound demos as well? What's the full Matlab output of PTB for a session? What sound hardware? Any warnings about not properly closing textures at the end of a session? Could be some out of memory condition although 12 Gb RAM make it somewhat less likely. Or not using an ASIO soundcard, or misconfigured sound, or some general Windows bug.

**Comment by alirezakzmi** [comment]
In the following, I provided the PTB output after each session. it contains the information of my hardware. However, I am not using ASIO sound card and I am fairly sure that the problem is not the sound player there should be something overloading that I cannot detect it appears on the sound and graphic both.

`
	PTB-INFO: Using specially modified PortAudio engine, based on offical version: PortAudio V19-devel WITH-DIM

	Will use ASIO enhanced Portaudio driver DLL. See Psychtoolbox/PsychSound/PortAudioLICENSE.txt
	for the exact terms of use for this dll.


	Disclaimer: "ASIO is a trademark and software of Steinberg Media Technologies GmbH."
	PTB-Warning: Although using the ASIO enabled Psychtoolbox sound driver,
	PTB-Warning: could not find any ASIO capable soundcard in your system.
	PTB-Warning: If you think you should have an ASIO card, please check your
	PTB-Warning: system for properly installed and configured drivers and retry.
	PTB-Warning: Read "help InitializePsychSound" for more info about ASIO et al.
	PTB-INFO: New audio device -1 with handle 0 opened as PortAudio stream:
	PTB-INFO: For 2 channels Playback: Audio subsystem is Windows DirectSound, Audio device name is Primary Sound Driver
	PTB-INFO: Real samplerate 44100.000000 Hz. Input latency 0.000000 msecs, Output latency 39.977324 msecs.


	Will use ASIO enhanced Portaudio driver DLL. See Psychtoolbox/PsychSound/PortAudioLICENSE.txt
	for the exact terms of use for this dll.


	Disclaimer: "ASIO is a trademark and software of Steinberg Media Technologies GmbH."
	PTB-Warning: Although using the ASIO enabled Psychtoolbox sound driver,
	PTB-Warning: could not find any ASIO capable soundcard in your system.
	PTB-Warning: If you think you should have an ASIO card, please check your
	PTB-Warning: system for properly installed and configured drivers and retry.
	PTB-Warning: Read "help InitializePsychSound" for more info about ASIO et al.
	PTB-INFO: New audio device -1 with handle 1 opened as PortAudio stream:
	PTB-INFO: For 2 channels Playback: Audio subsystem is Windows DirectSound, Audio device name is Primary Sound Driver
	PTB-INFO: Real samplerate 44100.000000 Hz. Input latency 0.000000 msecs, Output latency 39.977324 msecs.



	PTB-INFO: This is Psychtoolbox-3 for Microsoft Windows, under Matlab 64-Bit (Version 3.0.14 - Build date: Oct  3 2017).
	PTB-INFO: Support status on this operating system release: Windows version 10.0 supported and tested to some limited degree.
	PTB-INFO: Type 'PsychtoolboxVersion' for more detailed version information.
	PTB-INFO: Most parts of the Psychtoolbox distribution are licensed to you under terms of the MIT License, with
	PTB-INFO: some restrictions. See file 'License.txt' in the Psychtoolbox root folder for the exact licensing conditions.

	PTB-INFO: The detected endline of the vertical blank interval is equal or lower than the startline. This indicates
	PTB-INFO: that i couldn't detect the duration of the vertical blank interval and won't be able to correct timestamps
	PTB-INFO: for it. This will introduce a very small and constant offset (typically << 1 msec). Read 'help BeampositionQueries'
	PTB-INFO: for how to correct this, should you really require that last few microseconds of precision.
	PTB-INFO: Btw. this can also mean that your systems beamposition queries are slightly broken. It may help timing precision to
	PTB-INFO: enable the beamposition workaround, as explained in 'help ConserveVRAMSettings', section 'kPsychUseBeampositionQueryWorkaround'.


	PTB-INFO: OpenGL-Renderer is Intel :: Intel(R) HD Graphics 5300 :: 4.3.0 - Build 10.18.15.4256
	PTB-INFO: VBL startline = 1080 , VBL Endline = 1080
	PTB-INFO: Measured monitor refresh interval from beamposition = 16.681891 ms [59.945244 Hz].
	PTB-INFO: Will use beamposition query for accurate Flip time stamping.
	PTB-INFO: Measured monitor refresh interval from VBLsync = 16.684565 ms [59.935637 Hz]. (50 valid samples taken, stddev=0.069477 ms.)
	PTB-INFO: Reported monitor refresh interval from operating system = 16.949153 ms [59.000000 Hz].
	PTB-INFO: Small deviations between reported values are normal and no reason to worry.
	PTB-INFO: ==============================================================================================================================
	PTB-INFO: WINDOWS DWM DESKTOP COMPOSITOR IS ACTIVE. On this Windows-10 or later system, Psychtoolbox can no longer reliably detect if
	PTB-INFO: this will cause trouble for timing and integrity of visual stimuli or not. You might be just fine, or you could be in trouble.
	PTB-INFO: Use external measurement equipment and independent procedures to verify reliability of timing if you care about proper timing.
	PTB-INFO: ==============================================================================================================================


	PTB-INFO: This is Psychtoolbox-3 for Microsoft Windows, under Matlab 64-Bit (Version 3.0.14 - Build date: Oct  3 2017).
	PTB-INFO: Support status on this operating system release: Windows version 10.0 supported and tested to some limited degree.
	PTB-INFO: Type 'PsychtoolboxVersion' for more detailed version information.
	PTB-INFO: Most parts of the Psychtoolbox distribution are licensed to you under terms of the MIT License, with
	PTB-INFO: some restrictions. See file 'License.txt' in the Psychtoolbox root folder for the exact licensing conditions.

	PTB-INFO: The detected endline of the vertical blank interval is equal or lower than the startline. This indicates
	PTB-INFO: that i couldn't detect the duration of the vertical blank interval and won't be able to correct timestamps
	PTB-INFO: for it. This will introduce a very small and constant offset (typically << 1 msec). Read 'help BeampositionQueries'
	PTB-INFO: for how to correct this, should you really require that last few microseconds of precision.
	PTB-INFO: Btw. this can also mean that your systems beamposition queries are slightly broken. It may help timing precision to
	PTB-INFO: enable the beamposition workaround, as explained in 'help ConserveVRAMSettings', section 'kPsychUseBeampositionQueryWorkaround'.


	PTB-INFO: OpenGL-Renderer is Intel :: Intel(R) HD Graphics 5300 :: 4.3.0 - Build 10.18.15.4256
	PTB-INFO: VBL startline = 1080 , VBL Endline = 1080
	PTB-INFO: Measured monitor refresh interval from beamposition = 16.682040 ms [59.944706 Hz].
	PTB-INFO: Will use beamposition query for accurate Flip time stamping.
	PTB-INFO: Measured monitor refresh interval from VBLsync = 16.683846 ms [59.938218 Hz]. (50 valid samples taken, stddev=0.083440 ms.)
	PTB-INFO: Reported monitor refresh interval from operating system = 16.949153 ms [59.000000 Hz].
	PTB-INFO: Small deviations between reported values are normal and no reason to worry.
	PTB-INFO: ==============================================================================================================================
	PTB-INFO: WINDOWS DWM DESKTOP COMPOSITOR IS ACTIVE. On this Windows-10 or later system, Psychtoolbox can no longer reliably detect if
	PTB-INFO: this will cause trouble for timing and integrity of visual stimuli or not. You might be just fine, or you could be in trouble.
	PTB-INFO: Use external measurement equipment and independent procedures to verify reliability of timing if you care about proper timing.
	PTB-INFO: ==============================================================================================================================


	INFO: PTB's Screen('Flip', 10) command seems to have missed the requested stimulus presentation deadline
	INFO: a total of 304 times out of a total of 1566 flips during this session.

	INFO: This number is fairly accurate (and indicative of real timing problems in your own code or your system)
	INFO: if you provided requested stimulus onset times with the 'when' argument of Screen('Flip', window [, when]);
	INFO: If you called Screen('Flip', window); without the 'when' argument, this count is more of a ''mild'' indicator
	INFO: of timing behaviour than a hard reliable measurement. Large numbers may indicate problems and should at least
	INFO: deserve your closer attention. Cfe. 'help SyncTrouble', the FAQ section at www.psychtoolbox.org and the
	INFO: examples in the PDF presentation in PsychDocumentation/Psychtoolbox3-Slides.pdf for more info and timing tips.

`

**Comment by kleinerm** [comment]
On Windows anything but ASIO can be highly problematic if you only push it slightly. And audio timing is usually quite wrong on anything but ASIO hardware under Windows. So i don't know. You could try to provide the optional suggestedLatencySecs parameter in 'Open' with a really large latency of say 0.2 secs or more in the hope of working around fragility in how sound processing is scheduled. The defaults are about 40 msecs in "low latency mode" with reqlatencyclass > 0, and about 120 msecs in high latency mode reqlatencyclass = 0. Maybe 120 msecs are still too tight for your systems? Or try installing ASIO4All as an emulated ASIO driver, sometimes it helps.

Btw. NVidia Optimus graphics under Windows usually provides totally broken visual timing, so only using the Intel graphics is probably a good idea anyway in the cases where Intel gfx isn't also broken.

**Comment by adrofes** [comment]
Thank you for the answers. The latency did not matter much in my case so to avoid those lags, I indicated 0 in [,reqlatencyclass]. 

pahandle = PsychPortAudio('Open' [, deviceid][, mode]**[, reqlatencyclass]**[, freq][, channels][, buffersize][, suggestedLatency][, selectchannels][, specialFlags=0]);

**Comment by amir7697** [comment]
Hi!
Unfortunately, I have the same problem.
Did anyone find a way to solve this problem?

**Comment by kleinerm** [comment]
If you have the same problem, then the same solutions apply. If your problem is only similar then describe it in exact detail, with hardware/software specs, ptb output etc.

**Comment by gmarkkula** [comment]
Just commenting quickly here to say thanks for the advice above, and to add some further info for the benefit of any others having the same problem. I had this sound distortion on two PCs, one Win7 (my own computer) one Win10 (the lab computer), both with non-ASIO soundcards. Changing the reqlatencyclass  parameter as described in comments above made the sound play without distortion, but in my case the associated latency increase (and especially the increase in latency variability) wasn't great for what I was doing. However, installing ASIO4All fixed the problem completely on the Win10 machine, to my great relief. (I have not tried it on my own machine - no admin rights!)

**Comment by kleinerm** [comment]
@gmarkkula note that ASIO is no longer supported by past and current Psychtoolbox. Windows WASAPI is now used as sound system on Windows, starting with Psychtoolbox-3.0.15. It should hopefully have comparable performance.

**Comment by gmarkkula** [comment]
Ah thanks for sharing this info. This was indeed for slightly older 3.0.x versions.

**Comment by JuliaPsych** [comment]
Hi hope it's okay for me to continue this thread as my problem original started out as then one described above. I would play audio using PTB 3.0.14, running on Matlab 2016a, Windows 10. Sometimes audio would become distorted or sound robotic to participants who were rating audio stimuli. There was little load on the system as I loaded each stimulus at a time and timing precision was not relevant. 

After updating PTB to use the WASAPI rather than the old ASIO. However, I got the error that my requested sampling rate of 44100 with 1 channel were not supported by my hardware. Hence, I changed the device index for PsychPortAudio('Open') so that it uses the MME sound system because that was the only one that would work for me. Given that timing is not of critical interest here that's fine but I was wondering whether there might be a fix in the future with regards to WASAPI supporting other then the default sampling rate of 48000. 

Many thanks :-)
